packages/pigeon/lib/cpp_generator.dart

```diff
@@ -800,79 +818,70 @@ const flutter::StandardMessageCodec& ${api.name}::GetCodec() {
+      const String argumentListVariableName = 'encoded_api_arguments';
+      indent.write('flutter::EncodableValue $argumentListVariableName = ');
+      if (func.arguments.isEmpty) {
+        indent.addln('flutter::EncodableValue();');
+      } else {
+        indent.scoped('flutter::EncodableValue(flutter::EncodableList{', '});',
+            () {
+          for (final _HostNamedType arg in hostArgs) {
+            final String encodedArgument = _wrappedHostApiArgumentExpression(
+                root, arg.name, arg.originalType, arg.hostType);
+            indent.writeln('$encodedArgument,');
+          }
+        });
+      }
+
       indent.write(
-          '$channel->Send($sendArgument, [callback](const uint8_t* reply, size_t reply_size) ');
+          '$channel->Send($argumentListVariableName, [on_success, on_error](const uint8_t* reply, size_t reply_size) ');
       indent.scoped('{', '});', () {
+        final String successCallbackArgument;
         if (func.returnType.isVoid) {
-          indent.writeln('callback();');
+          successCallbackArgument = '';
         } else {
+          successCallbackArgument = 'return_value';
+          final String encodedReplyName = 'encodable_$successCallbackArgument';
           indent.writeln(
-              'std::unique_ptr<flutter::EncodableValue> decoded_reply = GetCodec().DecodeMessage(reply, reply_size);');
-          indent.writeln(
-              'flutter::EncodableValue args = *(flutter::EncodableValue*)(decoded_reply.release());');
-          const String output = 'output';
-
-          final bool isBuiltin =
-              _baseCppTypeForBuiltinDartType(func.returnType) != null;
-          final String returnTypeName =
-              _baseCppTypeForDartType(func.returnType);
-          if (func.returnType.isNullable) {
-            indent.writeln('$returnType $output{};');
-          } else {
-            indent.writeln('$returnTypeName $output{};');
-          }
-          const String pointerVariable = '${_pointerPrefix}_$output';
-          if (func.returnType.baseName == 'int') {
-            indent.format('''
-if (const int32_t* $pointerVariable = std::get_if<int32_t>(&args))
-\t$output = *$pointerVariable;
-else if (const int64_t* ${pointerVariable}_64 = std::get_if<int64_t>(&args))
-\t$output = *${pointerVariable}_64;''');
-          } else if (!isBuiltin) {
-            indent.write(
-                'if (const flutter::EncodableList* $pointerVariable = std::get_if<flutter::EncodableList>(&args)) ');
-            indent.scoped('{', '}', () {
-              indent.writeln('$output = $returnTypeName(*$pointerVariable);');
-            });
-          } else {
-            indent.write(
-                'if (const $returnTypeName* $pointerVariable = std::get_if<$returnTypeName>(&args)) ');
-            indent.scoped('{', '}', () {
-              indent.writeln('$output = *$pointerVariable;');
-            });
-          }
-
-          indent.writeln('callback($output);');
```

Comment:
This code was replaced with `_writeEncodedArgumentExtraction`, which was more comprehensive/correct.